src/pubsub/ua_pubsub_reader.c

```diff
@@ -287,7 +285,7 @@ void UA_ReaderGroup_subscribeCallback(UA_Server *server, UA_ReaderGroup *readerG
         UA_LOG_INFO(&server->config.logger, UA_LOGCATEGORY_SERVER, "Message buffer alloc failed!");
         return;
     }
-
+    buffer.length = 0;
```

Comment:
Not sure about that one.
Isn't the length of the available buffer checked in the `receive` function?

Reply:
buffer length is initialized to 0 here.  After the receive() function,  the length of the buffer is checked to determine if a message is received or not. The check can successful due to garbage value even if the message is not received.

Reply:
Upon closer look, this is not correct. The buffer.length is the total size of the buffer, not the size of the received buffer. It is used in memset(0) in allocate and delete buffer functions. 

But it is also used to check the length of the received message on the next line. In fact, the decision, whether or not a message is received,  is made by checking this length. Total allocated length is 512 bytes, so this check would always be true whether or not the message is received.  Currently, MQTT channel doesn't have receive function, code just explodes at this point. Both UDPMP and ETHERNET receive functions change the buffer length to 0 or to the received message size. This also causes memory leaks. We can maybe change the structure of  UA_ByteString, add one more variable for current size. In any case, the handling of pubsub receive must be change.

For now, I am removing this line.

Reply:
> Both UDPMP and ETHERNET receive functions change the buffer length to 0 or to the received message size. This also causes memory leaks.

That should not happen. The line that sets the length to zero should also free the memory.
Setting to a smaller positive value is okay as `free` does not care in the end.

Can you open separate issues if the above is confirmed?

Reply:
It is being freed at the end of subscriberCallBack() in all cases. I was concerned about the memset(0). Didn't realize that type->size and buffer size are different. 
All good!